kms/asymmetric/samples.go

```diff
@@ -0,0 +1,189 @@
+	if err != nil {
+		return "", err
+	}
+
+	//perform type assertion to get rsa key
+	rsaKey := abstractKey.(*rsa.PublicKey)
+
+	ciphertextBytes, err := rsa.EncryptOAEP(sha256.New(), rand.Reader, rsaKey, []byte(message), nil)
+	if err != nil {
+		return "", err
+	}
+	ciphertextStr := base64.StdEncoding.EncodeToString(ciphertextBytes)
+	return ciphertextStr, nil
+}
+
+// [END kms_encrypt_rsa]
+
+// [START kms_sign_asymmetric]
+// Sign a message using saved asymmetric private key
+func signAsymmetric(client *cloudkms.Service, ctx context.Context, message, keyPath string) (string, error) {
+	//find hash of message
+	digest := sha256.New()
+	digest.Write([]byte(message))
+	digestStr := base64.StdEncoding.EncodeToString(digest.Sum(nil))
+
+	asymmetricSignRequest := &cloudkms.AsymmetricSignRequest{
+		Digest: &cloudkms.Digest{
+			Sha256: digestStr,
+		},
+	}
+
+	response, err := client.Projects.Locations.KeyRings.CryptoKeys.CryptoKeyVersions.
+		AsymmetricSign(keyPath, asymmetricSignRequest).Context(ctx).Do()
+	if err != nil {
+		return "", err
+	}
+
+	return response.Signature, nil
+}
+
+// [END kms_sign_asymmetric]
+
+// [START kms_verify_signature_rsa]
+// Verify the cryptographic signature for a message signed with an RSA private key
+func verifySignatureRSA(client *cloudkms.Service, ctx context.Context, signature, message, keyPath string) error {
+	abstractKey, err := getAsymmetricPublicKey(client, ctx, keyPath)
+	if err != nil {
+		return err
+	}
+	//perform type assertion to get rsa key
+	rsaKey := abstractKey.(*rsa.PublicKey)
+	decodedSignature, _ := base64.StdEncoding.DecodeString(signature)
+
+	digest := sha256.New()
+	digest.Write([]byte(message))
+	hash := digest.Sum(nil)
+
+	pssOptions := rsa.PSSOptions{SaltLength: len(hash), Hash: crypto.SHA256}
+	err = rsa.VerifyPSS(rsaKey, crypto.SHA256, hash, decodedSignature, &pssOptions)
+	if err != nil {
+		return fmt.Errorf("verification failed")
+	}
+	return nil
+}
+
+// [END kms_verify_signature_rsa]
+
+// [START kms_verify_signature_ec]
+// Verify the cryptographic signature for a message signed with an Elliptic Curve private key
+func verifySignatureEC(client *cloudkms.Service, ctx context.Context, signature, message, keyPath string) error {
+	abstractKey, err := getAsymmetricPublicKey(client, ctx, keyPath)
+	if err != nil {
+		return err
+	}
+	//perform type assertion to get elliptic curve key
+	ecKey := abstractKey.(*ecdsa.PublicKey)
+	decodedSignature, err := base64.StdEncoding.DecodeString(signature)
+
+	var parsedSig struct{ R, S *big.Int }
+	rest, err := asn1.Unmarshal(decodedSignature, &parsedSig)
+	if err != nil || len(rest) != 0 {
+		return err
+	}
+
+	digest := sha256.New()
+	digest.Write([]byte(message))
+	hash := digest.Sum(nil)
+
+	if !ecdsa.Verify(ecKey, hash, parsedSig.R, parsedSig.S) {
+		return fmt.Errorf("verification failed")
+	}
+	return nil
+}
+
+// [END kms_verify_signature_ec]
+
+func main() {
+	project_id := "sanche-testing-project"
```

Comment:
An idea to float around is to make this a command-line flag or env var so it can be more plug & playable

Reply:
good catch. The main function won't be part of the samples, but I'll make sure to take out my hardcoded projectID